server/src-lib/Hasura/Server/Auth/JWT.hs

```diff
@@ -20,29 +27,61 @@ import           Data.ASN1.Encoding         (decodeASN1')
 import           Data.ASN1.Types            (ASN1 (End, IntVal, Start),
                                              ASN1ConstructionType (Sequence),
                                              fromASN1)
+import           Data.Int                   (Int64)
+import           Data.IORef                 (IORef, modifyIORef, readIORef)
+
 import           Data.List                  (find)
-import           Data.Time.Clock            (getCurrentTime)
+import           Data.Time.Clock            (diffUTCTime, getCurrentTime)
+import           Data.Time.Format           (defaultTimeLocale, parseTimeM)
+
+import           Hasura.Logging             (LogLevel (..), Logger)
 import           Hasura.Prelude
 import           Hasura.RQL.Types
+import           Hasura.Server.Logging      (JwkRefreshHttpError (..),
+                                             JwkRefreshLog (..))
 import           Hasura.Server.Utils        (accessKeyHeader, bsToTxt,
                                              userRoleHeader)
 
+import qualified Control.Concurrent         as C
 import qualified Data.Aeson                 as A
 import qualified Data.Aeson.Casing          as A
 import qualified Data.Aeson.TH              as A
-
 import qualified Data.ByteString.Lazy       as BL
 import qualified Data.ByteString.Lazy.Char8 as BLC
 import qualified Data.CaseInsensitive       as CI
 import qualified Data.HashMap.Strict        as Map
 import qualified Data.PEM                   as PEM
+import qualified Data.String.Conversions    as CS
 import qualified Data.Text                  as T
 import qualified Data.Text.Encoding         as TE
 import qualified Data.X509                  as X509
+import qualified Network.HTTP.Client        as HTTP
 import qualified Network.HTTP.Types         as HTTP
+import qualified Network.URI                as N
+import qualified Network.Wreq               as Wreq
 
 
-type RawJWT = BL.ByteString
+newtype RawJWT = RawJWT BL.ByteString
+
+data JWTConfig
+  = JWTConfig
+  { jcType     :: !T.Text
+  , jcKey      :: !(Maybe JWK)
```

Comment:
This type will allow both `jcKey` and `jcJwkUrl` to be Nothing which is not a valid `JWTConfig`. It should instead be `jcKey :: Either JWK N.URI`

Reply:
The json parsing should account for this.